plutus-executables/executables/uplc/Main.hs

```diff
@@ -268,15 +268,33 @@ plutusOpts = hsubparser $
 
 -- | Run the UPLC optimisations
 runOptimisations :: OptimiseOptions -> IO ()
-runOptimisations (OptimiseOptions inp ifmt outp ofmt mode cert) = do
+runOptimisations (OptimiseOptions inp ifmt outp ofmt mode mcert) = do
   prog <- readProgram ifmt inp :: IO (UplcProg SrcSpan)
   (simplified, simplificationTrace) <- PLC.runQuoteT $ do
     renamed <- PLC.rename prog
     let defaultBuiltinSemanticsVariant :: BuiltinSemanticsVariant PLC.DefaultFun
         defaultBuiltinSemanticsVariant = def
     UPLC.simplifyProgramWithTrace UPLC.defaultSimplifyOpts defaultBuiltinSemanticsVariant renamed
   writeProgram outp ofmt mode simplified
-  runCertifier cert simplificationTrace
+  case mcert of
+    Nothing   -> pure ()
+    Just cert -> execCertifier simplificationTrace cert
+  where
+    execCertifier simplificationTrace cert = do
+      result <- runCertifier $ mkCertifier simplificationTrace cert
+      case result of
+        Left err -> do
+          putStrLn $ prettyCertifierError err
+          case err of
+            InvalidCertificate    -> exitWith $ ExitFailure 1
+            InvalidCompilerOutput -> exitWith $ ExitFailure 2
+            ValidationError _     -> exitWith $ ExitFailure 3
+        Right certDir -> do
+          putStrLn
+            $ "The compilation was successfully certified.\nAgda certificate project written to "
```

Comment:
We're printing this to stdout which is weird. Let's take care of this in https://github.com/IntersectMBO/plutus-private/issues/1544. That's where we should improve the output and the format.

Reply:
Btw I'm not doing this in this PR because it means messing around with the `Executable` tests again and it's a mess (as pointed out by the TODO). These two issues (1544 and 1582) should take care of the mess.